tools/criticalstat.py

```diff
@@ -0,0 +1,335 @@
+    u64 ts;
+    int idle_skip;
+    int active;
+};
+
+struct data_t {
+    u64 time;
+    u64 stack_id;
+    u32 cpu;
+    u64 id;
+    u32 addrs[4];   /* indexed by addr_offs */
+    char comm[TASK_COMM_LEN];
+};
+
+BPF_STACK_TRACE(stack_traces, 1024);
+BPF_PERCPU_ARRAY(sts, struct start_data, 1);
+BPF_PERCPU_ARRAY(isidle, u64, 1);
+BPF_PERF_OUTPUT(events);
+
+TRACEPOINT_PROBE(power, cpu_idle)
+{
+    int idx = 0;
+    u64 val;
+    struct start_data *stdp, std;
+
+    stdp = sts.lookup(&idx);
+    bpf_probe_read(&std, sizeof(struct start_data), stdp);
+
+    // Mark active sections as that they should be skipped
+
+    // Handle the case CSenter, Ienter, CSexit, Iexit
+    // Handle the case CSenter, Ienter, Iexit, CSexit
+    if (std.active) {
+        std.idle_skip = 1;
+        sts.update(&idx, &std);
+    }
+
+    // Mark CPU as actively within idle or not.
+    if (args->state < 100) {
+        bpf_trace_printk(\"Setting idle\\n\");
+        val = 1;
+        isidle.update(&idx, &val);
+    } else {
+        bpf_trace_printk(\"Resetting idle\\n\");
+        val = 0;
+        isidle.update(&idx, &val);
+    }
+    return 0;
+}
+
+static int in_idle(void)
+{
+     u64 *idlep, idle;
+     int idx = 0;
+
+    // Skip event if we're in idle loop
+    idlep = isidle.lookup(&idx);
+    bpf_probe_read(&idle, sizeof(u64), idlep);
+
+    // Skip if we're in idle loop
+    if (idle)
+        return 1;
+    else
+        return 0;
+}
+
+static void reset_state(void)
+{
+    int idx = 0;
+    struct start_data s = {};
+
+    sts.update(&idx, &s);
+}
+
+TRACEPOINT_PROBE(preemptirq, TYPE_disable)
+{
+    int idx = 0;
+    struct start_data s;
+
+    // Handle the case Ienter, CSenter, CSexit, Iexit
+    // Handle the case Ienter, CSenter, Iexit, CSexit
+    if (in_idle()) {
+        bpf_trace_printk(\"disable: In idle, resetting\\n\");
+        reset_state();
+        return 0;
+    }
+
+    u64 ts = bpf_ktime_get_ns();
+
+    bpf_trace_printk(\"Entered new section, setting time to %lu\\n\",
+        (unsigned long)ts);
+    s.idle_skip = 0;
+    s.addr_offs[start_caller_off] = args->caller_offs;
+    s.addr_offs[start_parent_off] = args->parent_offs;
+    s.ts = ts;
+    s.active = 1;
+    bpf_trace_printk(\"Finished storing\\n\", (unsigned long)ts);
+
+    sts.update(&idx, &s);
+    return 0;
+}
+
+TRACEPOINT_PROBE(preemptirq, TYPE_enable)
+{
+    int idx = 0;
+    u64 start_ts, end_ts, diff;
+    struct start_data *stdp, std;
+
+    // Handle the case CSenter, Ienter, CSexit, Iexit
+    // Handle the case Ienter, CSenter, CSexit, Iexit
+    if (in_idle()) {
+        bpf_trace_printk(\"enable: In idle, resetting\\n\");
+        reset_state();
+        return 0;
+    }
+
+    bpf_trace_printk(\"enable: start lookup\\n\");
+
+    stdp = sts.lookup(&idx);
+    bpf_trace_printk(\"enable: start read\\n\");
+    bpf_probe_read(&std, sizeof(struct start_data), stdp);
+
+    // Handle the case Ienter, Csenter, Iexit, Csexit
+    if (!std.active) {
```

Comment:
Same as above: the call to `bpf_probe_read` is not needed here with the latest bcc. This can be simplified to `if (stdp && stdp->active)`, which also handles the case when `stdp` is null.

Reply:
@pchaigno  I wish that worked. The failure I am now seeing is with with the `sts.update` as well in the if clause.
So if I do:
```
    if (stdp && stdp->active) {
        stdp->idle_skip = 1;
        sts.update(&idx, stdp);
    }
```
I get:
```
./criticalstat.py 
bpf: Failed to load program: Permission denied
0: (bf) r6 = r1
1: (b7) r1 = 0
2: (63) *(u32 *)(r10 -4) = r1
3: (18) r1 = 0xffffffc0439ea000
5: (bf) r2 = r10
6: (07) r2 += -4
7: (85) call 1
8: (bf) r7 = r0
9: (bf) r1 = r10
10: (07) r1 += -32
11: (b7) r2 = 24
12: (bf) r3 = r7
13: (85) call 4
14: (15) if r7 == 0x0 goto pc+11
 R0=inv,min_value=0,max_value=0 R6=ctx R7=map_value(ks=4,vs=24,id=0),min_value=0,max_value=0 R10=fp
15: (61) r1 = *(u32 *)(r7 +20)
16: (15) if r1 == 0x0 goto pc+9
 R0=inv,min_value=0,max_value=0 R1=inv R6=ctx R7=map_value(ks=4,vs=24,id=0),min_value=0,max_value=0 R10=fp
17: (b7) r1 = 1
18: (63) *(u32 *)(r7 +16) = r1
19: (18) r1 = 0xffffffc0439ea000
21: (bf) r2 = r10
22: (07) r2 += -4
23: (bf) r3 = r7
24: (b7) r4 = 0
25: (85) call 2
R3 type=map_value expected=fp

Traceback (most recent call last):
  File "./criticalstat.py", line 252, in <module>
    b = BPF(text=bpf_text)
  File "/usr/lib/python2.7/dist-packages/bcc/__init__.py", line 328, in __init__
    self._trace_autoload()
  File "/usr/lib/python2.7/dist-packages/bcc/__init__.py", line 1014, in _trace_autoload
    fn = self.load_func(func_name, BPF.TRACEPOINT)
  File "/usr/lib/python2.7/dist-packages/bcc/__init__.py", line 368, in load_func
    (func_name, errstr))
Exception: Failed to load BPF program tracepoint__power__cpu_idle: Permission denied
```

Reply:
Tell you what. I'll finish the docs and post another PR. I would like wrap up the docs this week. Then we can discuss this issue there. Thanks!